### PR TITLE
Enable warning level 4 for Visual Studio

### PR DESCRIPTION
**Pull request by KineticTheory**
### Background

* Warning level 4 for Visual Studio adds static analysis warnings for potential design flaws like data and function hiding.  SQA sites recommend running VS with `/W4` because these warnings are quite useful.

### Description of changes

Implement minor changes to eliminate newly created warnings.

+ **Data hiding**: Rename some variables so they do not hide variables with the same name declared at a higher scope.
+ **Function hiding**: Use `override` to be explicit about what functions should be used in inheritance hierarchies.
+ Change some signed/unsigned integers to eliminate unequal comparisons.
+ Add a `LASTENTRY` to the `Meshes` enum in `TestRTT_Format_Reader.hh` that can be used to initialize variables at declaration in the tests.
+ Eliminate `using namespace` commands found outside of function or class definitions. Also remove `using namespace` commands that appear inside of the same namespace.
+ Use `static_cast` in a few places to be explicit about data type conversion.
+ Use `constexpr` in a few places to eliminate warnings about conditionals that are known at compile time.
+ Add extra pragmas to suppress warnings that eminate from TPL code.
+ Eliminate `rng/Subrandom_Generator.cc` by moving the constructor definition to the header file.

Also:

+ Split lines that declared multiple variables on a single line.
+ Add two new functions to `DracoStrings` and use these instead of locally defined functions.  The new functions are constructed in a way that avoids warnings about type conversion.
  - `string_tolower`
  - `string_toupper`
+ Comment out some unused variable names found in function signatures.
+ Comment out some unreachable code.
+ Fix some include directives.
+ Remove `DLL_PUBLIC` macros where they are no longer needed.
+ Remove some dead code.
+ Fix comments (indentation, limit to 80 cols)

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

**Comment by KineticTheory** [comment]
@RyanWollaeger This is a fairly expansive PR.  Do you think you might have time to review it over the next few days?

**Comment by RyanWollaeger** [comment]
@KineticTheory Certainly.

File: src/diagnostics/draco_info.cc [hunk 1]
```diff
@@ -26,7 +26,7 @@ namespace rtt_diagnostics {
 DracoInfo::DracoInfo(void)
     : release(rtt_dsxx::release()), copyright(rtt_dsxx::copyright()),
       contact("For information, send e-mail to draco@lanl.gov."),
-      build_type(normalizeCapitalization(CMAKE_BUILD_TYPE)),
+      build_type(rtt_dsxx::string_toupper(CMAKE_BUILD_TYPE)),
```
Comment (RyanWollaeger) [inline src/diagnostics/draco_info.cc]: Nice!

File: src/ds++/path_getFilenameComponent.cc [hunk 2]
```diff
@@ -116,10 +116,11 @@ std::string getFilenameComponent(std::string const &fqName,
...
 
   // Always convert paths to use native format.
-  if (dirSep == WinDirSep) // Windows style.
-    std::replace(retVal.begin(), retVal.end(), UnixDirSep, dirSep);
-  else
-    std::replace(retVal.begin(), retVal.end(), WinDirSep, dirSep);
+#ifdef WIN32
+  std::replace(retVal.begin(), retVal.end(), UnixDirSep, dirSep);
+#else
+  std::replace(retVal.begin(), retVal.end(), WinDirSep, dirSep);
```
Comment (RyanWollaeger) [inline src/ds++/path_getFilenameComponent.cc]: This conditional looks swapped to me - I would think it should be:
```
#ifdef WIN32
  std::replace(retVal.begin(), retVal.end(), WinDirSep, dirSep);
#else
  std::replace(retVal.begin(), retVal.end(), UnixDirSep, dirSep);
```
Apologies if I'm missing something.

File: src/rng/Subrandom_Generator.hh [hunk 3]
```diff
@@ -58,7 +58,13 @@ protected:
 
   // IMPLEMENTATION
 
-  Subrandom_Generator(unsigned count);
+  /*!
+   * \brief Constructor for Subrandom_Generator.
+   * \param count Dimension of the vector calculated by this generator.
```
Comment (RyanWollaeger) [inline src/rng/Subrandom_Generator.hh]: `\param` -> `\param[in]?`

File: src/ds++/path_getFilenameComponent.cc (hunk 2 above)
Comment (RyanWollaeger) [reply src/ds++/path_getFilenameComponent.cc]: Oh - apologies, I was confusing the old and new string in the `std::replace` function.

File: src/c4/test/tstProcessor_Group.cc [hunk 4]
```diff
@@ -69,21 +69,20 @@ void tstProcessor_Group(rtt_dsxx::UnitTest &ut) {
     for (size_t i = 0; i < vlen; ++i)
       myvec.push_back(pid * 1000.0 + i);
     vector<double> globalvec(group_pids * vlen);
+    Check(myvec.size() < UINT32_MAX);
     comm.assemble_vector(&myvec[0], &globalvec[0],
                          static_cast<unsigned>(myvec.size()));
 
-    if (globalvec.size() != group_pids * vlen)
-      ITFAILS;
+    FAIL_IF_NOT(globalvec.size() == group_pids * vlen);
```
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/c4/test/tstProcessor_Group.cc]: I think so.  @timmah  added the `FAIL_IF_NOT` signature. I think the resulting code is shorter, cleaner and easier to read.

File: src/cdi_analytic/Analytic_Models.hh [hunk 5]
```diff
@@ -184,16 +184,16 @@ private:
...
+   * \param b_ temperature multiplier [keV^(-c) cm^2/g (cm^3/g)^d]
+   * \param c_ temperature power
+   * \param d_ density power
+   * \param e_ frequency power
+   * \param f_ reference temperature
+   * \param g_ reference density
+   * \param h_ reference frequency
+   */
 
   Polynomial_Analytic_Opacity_Model(double a_, double b_, double c_, double d_,
```
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/cdi_analytic/Analytic_Models.hh]: I think this is outside of the scope of this PR. @kgbudge and I are working and an update to the `cdi_analytic` package. See #472.

File: src/ds++/path_getFilenameComponent.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/ds++/path_getFilenameComponent.cc]: I've been looking into the differences between `WIN32` and `MSVC`.  For our target builds, there is no difference, but the `WIN32` CPP would also apply to a MSYS/MinGW or a Cygwin build, while `MSVC` is limited to the Microsoft compiler.

File: config/dracoVersion.cmake [hunk 6]
```diff
@@ -9,11 +9,11 @@
 
 macro( set_ccs2_software_version PROJNAME )
 
-  if( NOT ${${PROJNAME}_VERSION_MAJOR} )
+  if( "${${PROJNAME}_VERSION_MAJOR}x" STREQUAL "x" )
```
[the remark this answers is not in the document]
Comment (KineticTheory) [reply config/dracoVersion.cmake]: I don't like the way cmake handles empty variables.  I think the old coding should work, but it failed.  In my testing, I did the following:

```
set( temp2 "" )   # set temp2 to an empty string
if( NOT temp2 )
  message("temp2 not found")
endif()
if( NOT ${temp2} )
  message("evaluated temp2 not found")
endif()
if( NOT "${temp2}" )
  message("evaluated & quoted temp2 not found")
endif()
```

The result is

```
temp2 not found
evaluated & quoted temp2 not found
```

So, I think I could change the line in question to

```diff
-if( NOT ${${PROJNAME}_VERSION_MAJOR} )
+if( NOT "${${PROJNAME}_VERSION_MAJOR}" )
```

or 

Either way, the old check is broken.  I'll change this line to the last option listed in this comment.

File: config/vendor_libraries.cmake [hunk 7]
```diff
@@ -479,11 +479,23 @@ endmacro()
 #------------------------------------------------------------------------------
 macro( setupParMETIS )
 
+  set( QUIET "QUIET")
   if( NOT TARGET METIS::metis )
     message( STATUS "Looking for METIS..." )
 
-    find_package( METIS QUIET )
-    if( METIS_FOUND )
+    find_package( METIS CONFIG ${QUIET} )
```
[the remark this answers is not in the document]
Comment (KineticTheory) [reply config/vendor_libraries.cmake]: I'm going to revisit the ParMETIS `find_package` soon.  I'll postpone this change until I look more carefully at ParMETIS.

File: src/cdi_analytic/Analytic_Models.hh [hunk 8]
```diff
@@ -184,16 +184,16 @@ private:
...
-     * \param b_ temperature multiplier [keV^(-c) cm^2/g (cm^3/g)^d]
-     * \param c_ temperature power
-     * \param d_ density power
-     * \param e_ frequency power
-     * \param f_ reference temperature
-     * \param g_ reference density
-     * \param h_ reference frequency
-     */
+   * \brief Constructor.
+   * \param a_ constant [cm^2/g (cm^3/g)^d]
```
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/cdi_analytic/Analytic_Models.hh]: I'll add those.

File: src/rng/Subrandom_Generator.hh (hunk 3 above)
Comment (KineticTheory) [reply src/rng/Subrandom_Generator.hh]: I'll add this.

File: src/ds++/path_getFilenameComponent.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/ds++/path_getFilenameComponent.cc]: I'll defer action on this to issue #500.

File: config/dracoVersion.cmake (hunk 6 above)
[the remark this answers is not in the document]
Comment (KineticTheory) [reply config/dracoVersion.cmake]: Fixed in fe31d31

File: src/cdi_analytic/Analytic_Models.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/cdi_analytic/Analytic_Models.hh]: Fixed in fe31d31

File: src/cdi_analytic/Analytic_Models.hh (hunk 8 above)
[the remark this answers is not in the document]
Comment (KineticTheory) [reply src/cdi_analytic/Analytic_Models.hh]: Fixed in fe31d31

File: src/rng/Subrandom_Generator.hh (hunk 3 above)
Comment (KineticTheory) [reply src/rng/Subrandom_Generator.hh]: Fixed in fe31d31

**Pull request merged**